lib/export/dwca/data.rb

```diff
@@ -629,37 +887,664 @@ def biological_associations_resource_relationship_tmp
+              next
+            end
+
+            access_uri = Shared::Api.sound_link(snd)
+            further_info_url = Shared::Api.sound_metadata_link(snd, raise_on_no_token: true, token: token)
+
+            # NOTE: Removed associated_specimen_reference - now computed in SQL per row
+            links_data << {
+              sound_id: snd.id,
+              access_uri: access_uri,
+              further_information_url: further_info_url
+            }
+          rescue => e
+            Rails.logger.warn "dwca_export: skipping sound #{snd.id} - #{e.message}"
+          end
+        end
+
+        # Bulk insert sound links data
+        unless links_data.empty?
+          values = links_data.map do |row|
+            "(#{row[:sound_id]}, #{conn.quote(row[:access_uri])}, #{conn.quote(row[:further_information_url])})"
+          end.join(', ')
+
+          conn.execute("INSERT INTO temp_media_sound_links (sound_id, access_uri, further_information_url) VALUES #{values}")
+        end
+      end
+
+      Rails.logger.debug 'dwca_export: temp tables created with API links'
+    end
+
+    # Exports image media records to the output file using PostgreSQL COPY TO
+    # Streams data directly to avoid loading entire dataset into memory
+    def export_images_to_file(image_ids, output_file)
+      return if image_ids.empty?
+
+      conn = ActiveRecord::Base.connection
+      license_case = creative_commons_license_case_sql
+      copyright_label = copyright_label_sql
+
+      image_copy_sql = <<-SQL
+        COPY (
+          SELECT
+            REGEXP_REPLACE(dwc.\"occurrenceID\", E'[\\n\\t]', ' ', 'g') AS coreid,
+            REGEXP_REPLACE(
+              CASE
+                WHEN uuid_id.identifier IS NOT NULL THEN 'image:' || uuid_id.identifier
+                WHEN uri_id.identifier IS NOT NULL THEN 'image:' || uri_id.identifier
+                ELSE 'image:' || img.id::text
+              END,
+              E'[\\n\\t]', ' ', 'g'
+            ) AS identifier,
+            'Image' AS \"dc:type\",
+            img.id AS \"providerManagedID\",
+            #{license_case} AS \"dc:rights\",
+            #{license_case} AS \"dcterms:rights\",
+            REGEXP_REPLACE(owners.names, E'[\\n\\t]', ' ', 'g') AS \"Owner\",
+            NULL AS \"UsageTerms\",
+            REGEXP_REPLACE(#{copyright_label}, E'[\\n\\t]', ' ', 'g') AS \"Credit\",
+            REGEXP_REPLACE(creators.names, E'[\\n\\t]', ' ', 'g') AS \"dc:creator\",
+            REGEXP_REPLACE(creator_ids.identifiers, E'[\\n\\t]', ' ', 'g') AS \"dcterms:creator\",
+            REGEXP_REPLACE(dep.figure_label, E'[\\n\\t]', ' ', 'g') AS description,
+            REGEXP_REPLACE(dep.caption, E'[\\n\\t]', ' ', 'g') AS caption,
+            -- Compute associatedSpecimenReference directly from collection object ID for this row
+            CASE
+              WHEN co.id IS NOT NULL THEN '#{Shared::Api.host}/api/v1/collection_objects/' || co.id
```

Comment:
Hard-coding API endpoints too fragile, at minimum we need a constant to reference them.